From 85ae0798663467ba2ffa3fa37dde17f096a85f1d Mon Sep 17 00:00:00 2001
From: Markus Pfau <pfau@peak3d.de>
Date: Sat, 14 Jan 2017 17:20:38 +0100
Subject: [PATCH 1/1] fix mp4 keyframe_pts_only

---
 drivers/amlogic/amports/vmpeg4.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/amlogic/amports/vmpeg4.c b/drivers/amlogic/amports/vmpeg4.c
index 794d3c8..403cd04 100644
--- a/drivers/amlogic/amports/vmpeg4.c
+++ b/drivers/amlogic/amports/vmpeg4.c
@@ -348,7 +348,7 @@ static irqreturn_t vmpeg4_isr(int irq, void *dev_id)
 		}
 
 		if ((I_PICTURE == picture_type) ||
-				((P_PICTURE == picture_type) && (keyframe_pts_only != 0))) {
+				((P_PICTURE == picture_type) && (keyframe_pts_only == 0))) {
 			offset = READ_VREG(MP4_OFFSET_REG);
 			/*2500-->3000,because some mpeg4
 			video may checkout failed;
-- 
2.1.4

