diff -Naur linux-6eaacba.orig/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c linux-6eaacba/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
--- linux-6eaacba.orig/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c	2017-01-16 11:54:25.819382202 +0100
+++ linux-6eaacba/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c	2017-01-16 11:57:41.221026035 +0100
@@ -2065,7 +2065,9 @@
 		priv->dma_cap.atime_stamp =
 		    (hw_cap & DMA_HW_FEAT_TSVER2SEL) >> 13;
 		/* 802.3az - Energy-Efficient Ethernet (EEE) */
-		priv->dma_cap.eee = (hw_cap & DMA_HW_FEAT_EEESEL) >> 14;
+		//priv->dma_cap.eee = (hw_cap & DMA_HW_FEAT_EEESEL) >> 14;
+		//disable eee
+		priv->dma_cap.eee = 0;
 		priv->dma_cap.av = (hw_cap & DMA_HW_FEAT_AVSEL) >> 15;
 		/* TX and RX csum */
 		priv->dma_cap.tx_coe = (hw_cap & DMA_HW_FEAT_TXCOESEL) >> 16;
