diff --git a/drivers/gpu/arm/ump/Kbuild b/drivers/gpu/arm/ump/Kbuild
index d58ede2c9c0..98765c5b80f 100644
--- a/drivers/gpu/arm/ump/Kbuild
+++ b/drivers/gpu/arm/ump/Kbuild
@@ -86,11 +86,6 @@ ump-y = common/ump_kernel_common.o \
 	linux/ump_osk_low_level_mem.o \
 	linux/ump_osk_misc.o \
 	linux/ump_kernel_random_mapping.o \
-	$(UDD_FILE_PREFIX)linux/mali_osk_atomics.o \
-	$(UDD_FILE_PREFIX)linux/mali_osk_locks.o \
-	$(UDD_FILE_PREFIX)linux/mali_osk_memory.o \
-	$(UDD_FILE_PREFIX)linux/mali_osk_math.o \
-	$(UDD_FILE_PREFIX)linux/mali_osk_misc.o
 
 
 
diff --git a/drivers/gpu/arm/ump/linux/ump_kernel_linux.c b/drivers/gpu/arm/ump/linux/ump_kernel_linux.c
index 75446618c78..833b91c2617 100644
--- a/drivers/gpu/arm/ump/linux/ump_kernel_linux.c
+++ b/drivers/gpu/arm/ump/linux/ump_kernel_linux.c
@@ -356,30 +356,6 @@ static int ump_file_ioctl(struct inode *inode, struct file *filp, unsigned int c
 	return err;
 }
 
-int map_errcode(_mali_osk_errcode_t err)
-{
-	switch (err) {
-	case _MALI_OSK_ERR_OK :
-		return 0;
-	case _MALI_OSK_ERR_FAULT:
-		return -EFAULT;
-	case _MALI_OSK_ERR_INVALID_FUNC:
-		return -ENOTTY;
-	case _MALI_OSK_ERR_INVALID_ARGS:
-		return -EINVAL;
-	case _MALI_OSK_ERR_NOMEM:
-		return -ENOMEM;
-	case _MALI_OSK_ERR_TIMEOUT:
-		return -ETIMEDOUT;
-	case _MALI_OSK_ERR_RESTARTSYSCALL:
-		return -ERESTARTSYS;
-	case _MALI_OSK_ERR_ITEM_NOT_FOUND:
-		return -ENOENT;
-	default:
-		return -EFAULT;
-	}
-}
-
 /*
  * Handle from OS to map specified virtual memory to specified UMP memory.
  */
